metal/resource_metal_device.go

```diff
@@ -348,8 +350,66 @@ func resourceMetalDevice() *schema.Resource {
 				Default:     false,
 				ForceNew:    false,
 			},
+			"reinstall": {
+				Type:     schema.TypeList,
+				MaxItems: 1,
+				Optional: true,
+				Elem: &schema.Resource{
+					Schema: map[string]*schema.Schema{
+						"enabled": {
+							Type:        schema.TypeBool,
+							Description: "Whether the device should be reinstalled instead of destroyed",
+							Optional:    true,
+							Default:     false,
+						},
+						"deprovision_fast": {
+							Type:        schema.TypeBool,
+							Description: "Whether the OS disk should be filled with `00h` bytes before reinstall",
+							Optional:    true,
+							Default:     false,
+						},
+						"preserve_data": {
+							Type:        schema.TypeBool,
+							Description: "Whether the non-OS disks should be kept or wiped during reinstall",
+							Optional:    true,
+							Default:     false,
+						},
+					},
+				},
+			},
 		},
+		CustomizeDiff: customdiff.Sequence(
+			customdiff.ForceNewIf("custom_data", shouldReinstall),
```

Comment:
Nice condensing!

I wonder if `shouldReinstall` needs to take the field as an argument so it can check the that `custom_data` has infact changed before acting on it based on `reinstall`.

I don't know if `CustomizeDiff` is only triggered in the presence of a change in the named field or if it is always called, allowing the provider to use external factors to report a diff.

Reply:
func shouldReinstall(field string) func (_ context.Context, d *schema.ResourceDiff, meta interface{}) bool {
 return func (_ context.Context, d *schema.ResourceDiff, meta interface{}) bool {
    if ! diff on field return false
    ... do the existing function 
  }
}

Reply:
According to the docs:

> The return value of the condition function is ignored if the old and new values of the field compare equal, since no attribute diff is generated in that case.

https://pkg.go.dev/github.com/hashicorp/terraform-plugin-sdk/helper/customdiff?utm_source=godoc#ForceNewIf